File: Leetcode/Linked-list/160_intersection-of-two-linked-lists/main.cpp
// 160. 相交链表
// Created by bob on 2022/3/11.
//
#include <unordered_set>
#include "KStruct.h"

using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * 方法一：哈希集合
     */
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        unordered_set<ListNode *> visited;
        ListNode *temp = headA;
        while (temp != nullptr) {
            visited.insert(temp);
            temp = temp->next;
        }
        temp = headB;
        while (temp != nullptr) {
            if (visited.count(temp)){
                return temp;
            }
            temp = temp->next;
        }
        return nullptr;
    }

    /**
     * 方法二：双指针法
     * @param headA
     * @param headB
     * @return
     */
ListNode *getIntersectionNode2(ListNode *headA, ListNode *headB) {
    if (headA == nullptr || headB == nullptr) return nullptr;
    ListNode *tempA = headA;
    ListNode *tempB = headB;
    while (tempA != tempB) {
        tempA = tempA == nullptr ? headA : tempA->next;
        tempB = tempB == nullptr ? headB : tempB->next;
    }
    return tempA;
}
};

int main() {
    Solution s = Solution();
    auto *crossList = new ListNode(8, new ListNode(4, new ListNode(5)));
    auto *listA = new ListNode(4, new ListNode(1, crossList));
    auto *listB = new ListNode(5, new ListNode(6, new ListNode(1, crossList)));
    ListNode *result = s.getIntersectionNode2(listA, listB);
    printf("%d", result->val);  // expect: 8
}

File: Leetcode/Linked-list/160_intersection-of-two-linked-lists/main.py
from PyUtils.kstruct import ListNode


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    """双指针法"""
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None
        tempA = headA
        tempB = headB
        while tempA != tempB:
            tempA = tempB if tempA is None else tempA.next
            tempB = tempA if tempB is None else tempB.next
        return tempA


if __name__ == '__main__':
    cross_list = ListNode(8, ListNode(4, ListNode(5)))
    listA = ListNode(4, ListNode(1, cross_list))
    listB = ListNode(5, ListNode(6, ListNode(1, cross_list)))
    s = Solution()
    node = s.getIntersectionNode(listA, listB)
    print(node.val)  # expect: 8
